group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/InMemoryPartitionWriter.java

```diff
@@ -33,10 +37,141 @@
  */
 public class InMemoryPartitionWriter<T> implements PartitionWriter<T> {
 
+    public static class LogEntry {
+        public static <T> LogEntry value(T value) {
+            return new LogValue<>(value);
+        }
+
+        public static <T> LogEntry value(
+            long producerId,
+            short producerEpoch,
+            T value
+        ) {
+            return new LogValue<>(
+                producerId,
+                producerEpoch,
+                value
+            );
+        }
+
+        public static LogEntry control(
+            long producerId,
+            short producerEpoch,
+            int coordinatorEpoch,
+            TransactionResult result
+        ) {
+            return new LogControl(
+                producerId,
+                producerEpoch,
+                coordinatorEpoch,
+                result
+            );
+        }
+    }
+
+    public static class LogValue<T> extends LogEntry {
+        public final long producerId;
+        public final short producerEpoch;
+        public final T value;
+
+        private LogValue(
+            long producerId,
+            short producerEpoch,
+            T value
+        ) {
+            this.producerId = producerId;
+            this.producerEpoch = producerEpoch;
+            this.value = value;
+        }
+
+        private LogValue(T value) {
+            this(
+                RecordBatch.NO_PRODUCER_ID,
+                RecordBatch.NO_PRODUCER_EPOCH,
+                value
+            );
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
```

Comment:
Does this simply check that this object is of LogValue<?> type?

Reply:
Yes.